server_src/runners/history.js

```diff
@@ -25,15 +23,284 @@ const script = require("./scriptCheck.js");
+
+      const selectedFilament = null;
+      if (
+        printer.selectedFilament !== null &&
+        Array.isArray(printer.selectedFilament)
+      ) {
+        let profileId = [];
+        printer.selectedFilament.forEach((spool, index) => {
+          if (spool !== null) {
+            if (serverSettings[0].filamentManager) {
+              profileId = _.findIndex(profiles, function (o) {
+                return (
+                  o.profile.index ==
+                  printer.selectedFilament[index].spools.profile
+                );
+              });
+            } else {
+              profileId = _.findIndex(profiles, function (o) {
+                return o._id == printer.selectedFilament[index].spools.profile;
+              });
+            }
+            printer.selectedFilament[index].spools.profile =
+              profiles[profileId].profile;
+          }
+        });
+      }
+      if (historyCollection.length === 0) {
+        counter = 0;
+      } else {
+        counter =
+          historyCollection[historyCollection.length - 1].printHistory
+            .historyIndex + 1;
+      }
+
+      const printHistory = {
+        historyIndex: counter,
+        printerName: name,
+        printerID: printer._id,
+        costSettings: printer.costSettings,
+        success: true,
+        reason: payload.reason,
+        fileName: payload.name,
+        fileDisplay: payload.display,
+        filePath: payload.path,
+        startDate,
+        endDate,
+        thumbnail: "",
+        printTime: Math.round(payload.time),
+        filamentSelection: printer.selectedFilament,
+        previousFilamentSelection: previousFilament,
+        job,
+        notes: "",
+        snapshot: "",
+        resends: resends,
+      };
+
+      HistoryCollection.updateFilamentInfluxDB(
+        printer.selectedFilament,
+        printHistory,
+        previousFilament,
+        printer
+      );
+
+      const saveHistory = new History({
+        printHistory,
+      });
+      await saveHistory.save().then(async (r) => {
+        if (printer.camURL !== "") {
+          const thumbnail = await HistoryCollection.thumbnailCheck(
+            payload,
+            serverSettings[0],
+            files,
+            saveHistory._id,
+            "onFailure",
+            printer
+          );
+          const snapshot = await HistoryCollection.snapshotCheck(
+            "onFailure",
+            serverSettings[0],
+            printer,
+            saveHistory,
+            payload
+          );
+          if (
+            typeof serverSettings[0] !== "undefined" &&
+            typeof serverSettings[0].history !== "undefined" &&
+            typeof serverSettings[0].history.timelapse !== "undefined" &&
+            typeof serverSettings[0].history.timelapse.onComplete !==
+              "undefined" &&
+            serverSettings[0].history.timelapse.onComplete
+          ) {
+            HistoryCollection.timelapseCheck(
```

Comment:
At this point in the code we are indented like 6 tabs? Thats where I would say 'ok I gotta split up into functions'. As we are then nesting code into code making it hard to read. I try to reduce indentation to about max 3.
Whats your thought on that?

Reply:
I'd prefer, there's a hell of a lot of errant code that's getting nuked from this still though.